# SiCeLoRe

 
Acronyme for [Si]ngle [Ce]ll [Lo]ng [Re]ad is a suite of tools dedicated to the cell barcode and UMI (unique molecular identifier) assignment and bioinformatics analysis of highly multiplexed single cell Nanopore long read sequencing data.
Typically starting with a short read bam file from the 10xGenomics single cell cellranger pipeline and Nanopore long reads, the workflow integrates several sequential steps for cell barcode and UMI assignment to Nanopore reads (guided by Illumina data), transcript isoform identification, generation of transcript isoform count matrices and generation of consensus sequences (UMI-guided error-correction) for RNA molecules (UMIs).
[![GitHub license]()]((https://github.com/ucagenomix/sicelore2/blob/master/LICENCE.md))

 

# Installation

just copy files.

requires:

* Java 8,

* [fastp]([https://github.com/OpenGene/fastp)

* [Minimap2](https://github.com/lh3/minimap2),

* [Drop-seq tools v1.13](http://mccarrolllab.com/download/1276/),

* [poa](https://github.com/tanghaibao/bio-pipeline/tree/master/poaV2)

* [racon](https://github.com/isovic/racon)

* samtools

 

# Worflow

<img src="flow.png">


# Features

 

* [Parsing of Illumina Data](#parsing-illumina-data)   

    + Parses short read data and retrieves info on used cell barcodes and UMIs.

* [Nanopore poly(A) scan - stranding of reads](#nanopore-scan)  

    + Pre-scan of Nanopore reads for poly(A) tails -> stranded reads.

* [Mapping of Nanopore reads to the reference genome with minimap2](#minimap2-mapping)  

* [Tag Nanopore SAM records with gene names, read sequence and quality values](#nanopore-tagging)  

    + Adds gene names, read sequence and QV values. Required for barcode and UMI assignment

* [Barcode and UMI assignment to Nanopore SAM records](#bc-assignment)  

     

* [Transcript isoform expression quantification](#IsoformMatrix)  

    + Identifies matching Refseq transcript isoforms and generates isoform count table.

* [Consensus sequence calculation for RNA molecules (UMIs)](#ComputeConsensus)  

    + Generates consensus sequence for transcripts from multiple reads for UMI. 

* [Re-mapping of consensus for RNA molecules to the genome](#molecule-consensus-mapping)  

    + Consensus sequences are re-mapped to the reference genome

 

## Authors

 

* Kevin Lebrigand <lebrigand@ipmc.cnrs.fr> 

![Twitter Follow](https://img.shields.io/twitter/follow/kevinlebrigand.svg?style=social&logo=twitter)](https://twitter.com/kevinlebrigand)  

* Rainer Waldmann <waldmann@ipmc.cnrs.fr> 

 

<a id="parsing-illumina-data"></a>

## 1) Parsing of Illumina Data

 
Genome mapped short read data generated by the 10xGenomics CellRanger software (typically "possorted_genome_bam.bam" file) are parsed and info on cell barcodes and UMIs associated with each gene or genomic region are saved in a serialized nested Java Hashtable  which is required for barcode and UMI assignment to Nanopore reads.

 

### Required files

 
* IlluminaParser.jar

* Libraries in the ./lib folder  

* genome matched, barcode and UMI assigned Illumina short read data (e.g. bam file generated by 10x genomics cell ranger). 

* the 10xGenomics CellRanger tsv file that contains list of cell associated barcodes (available as barcodes.tsv file in outs/filtered_feature_bc_matrix directory of cellranger count analysis) or tsv file formatted the same  way if another system single cell library preparation was used.  

### Usage 

 

```bash

java  -Xmx15000m -jar IlluminaParser.jar --inFileIllumina /share/data/analysis/000-10x-cellranger/10x_waldmann/RWbrain200_2runs/outs/possorted_genome_bam.bam 

--tsv /share/data/analysis/000-10x-cellranger/10x_waldmann/RWbrain200_2runs/outs/filtered_gene_bc_matrices/mm10/barcodes.tsv 

--outFile /share/data/analysis/000-10x-cellranger/10x_waldmann/RWbrain200_2runs/outs/parsedForNanopore_v0.2.obj --cellBCflag CB --umiFlag UB --geneFlag GN


```


### Parameters  

 

**-b,--cellBCflag** (required)  

SAM tag for cell BC in Illumina bam file. Cell barcodes in cellranger bam files have a "-1" at the end. If other single cell sequencing systems were used, the "-1" at the end of the BC is not required. The "-" and following characters are ignored.   This is the assigned cell Barcode and not the read sequence for the cell barcodeI. In Cell Ranger bam files it is the BC tag. 

**-u,--umiFlag** (required)  

SAM tag for umi in Illumina bam file. This is the assigned UMI and not the read sequence for the UMI. In Cell Ranger bam files this is the sequence in the UB tag.  

**-g,--geneFlag** (required)  

SAM tag for Gene name in Illumina bam file.   

 **-i,--inFileIllumina** (required)  

path of bam file with genome matched Illumina data generated by 10xGenomics CellRanger  

**-t,--tsv** (required)  

the 10xGenomics tsv file that contains list of cell associated barcodes.   The file contains the list of cell barcodes that are associated with a cell. One cell barcode per line. 10x Genomics barcode tsv files  have a "-1" appended to the barcode sequence. The "-1" is not required and can be omitted if non-10xGenomics systems are used.   

**-o,--outFile** (required)  

full path of output file where the Illumina barcode/UMI data are saved.  File required for [Barcode and UMI assignment to Nanopore SAM records](#bc-assignment) .

 

<a id="nanopore-scan"></a>

## 2) Scan for poly(A) and adapters in Nanopore reads. 

 

Scans the Nanopore fastq reads for poly(A/T) and adapter sequence and generates stranded (forward) reads for reads with found  polyA and adapter.   

Scans by default for >= 15 nt. polyA (or T)  with >= 75% As within 100 nt from both ends of the read. If poly(A) was found, Searches for a 10xGenomics adapter sequence "CTTCCGATCT" downstream of the poly(A). 

When poly(A) and adapter were found at one end the read is  written stranded (forward) into a "pass" folder.  

Failed reads are written unstranded into a "failed" folder.

  

This is an optional step. Cell barcode and UMI assignment also works with non-stranded records. 

 

### Required files 

 

* NanoporeReadScanner.jar  

* Libraries in the ./lib folder  

* Config file: ReadScannerConfig.xml (Most default settings can be changed there).  

If no ReadScannerConfig.xml file is found in the current path (working directory), the software takes the default config file from the directory where the applications (jars)  are installed.

 

### Usage

 

```bash 

java -jar <path>/NanoporeReadScanner.jar -d <directory to start recursive search for fastq files>

```

 

### Parameters  

**-d,--inDir** (either this or –fastqFiles required)  

directory to start file search. starting at this directory, takes recursively fastq files that match the RegEx pattern given in –pattern  

**-i,--fastqFiles** (either this or – inDir required)  

" ,"  seperated list of fastq files  

**-v,-- pattern** (optional, defaults)  

fastq File name pattern to search when parsing (recursively) folders: defaults to: "**.{1,}\.fastq**"  

**-f,--fractionAT** (optional, defaults)  

min fraction AT, defaults to value in ReadScannerConfig.xml (0.75)  

**-p,--polyAlength** (optional, defaults)  

min length of polyA, defaults to 15  

**-w,--windowAT** (optional, defaults)  

window to search for  AT from the extremities of the read, defaults to 100 nt.  

**-o,--outDir**  

Output Directory, Creates a “failed” and “passed” subfolder with failed and stranded passed reads (found polyA and adapter) respectively.  

If <null> is given as output directory it won't write but just generate some stats.   

 

<a id="minimap2-mapping"></a>

## 3) Mapping of Nanopore reads to the reference genome with minimap2

 

### fastq splitting into chunks for paralellization  

 

Can be omitted for small runs (< 5 million reads)  

Prior to mapping the fastqs are split into chunks.  

uses [fastp]([https://github.com/OpenGene/fastp)  

Mapping, [tagging with gene names](#nanopore-tagging) and [Barcode/UMI assignment](#bc-assignment) are sequentially done with those batches to allow paralellization.

 

```console   

fastp -i PROMxxxxxx.fastq -Q -A --thread 20 --split_prefix_digits=4 --out1=sub.fastq --split=24

```

 

### parallel minimap2 mapping   

 

command shown for fastq batch "0001.sub.fastq"  

 

```bash  

minimap2 -ax splice -uf --MD -N 100 --sam-hit-only -t 20 --junc-bed gencode.vM18.primary_assembly.annotation.bed $BUILD.mmi 0001.sub.fastq > 0001.sub.sam

samtools view -Sb 0001.sub.sam -o 0001.sub.unsorted.bam

samtools sort 0001.sub.unsorted.bam -o 0001.sub.bam

samtools index 0001.sub.bam

```

**--junc-bed** (required)  

BED file consisting of annotated introns and their strands. With this option, minimap2 prefers splicing in annotations.

can be generated with `paftools.js gff2bed -j ann.gtf' 


<a id="nanopore-tagging"></a>

## 4) Tag Nanopore SAM records with gene names, read sequence and quality values 

 

### add gene names to Nanopore SAM records  

Adds gene names to Nanopore  SAMrecords  GE tag using [Drop-seq tools v1.13](http://mccarrolllab.com/download/1276/).

If another tag is used for gene name. Gene name SAM tag needs to be changed using TAG argument in dropseq.jar command line.

Fast, all jobs can be run on one compute node or distributed on a cluster. 

 
uses **TagReadWithGeneExon**  

command shown is for BAM batch "0001.sub.bam"  

 

```bash

cd ~/Drop-seq_tools-1.13/jar/

java -jar -Xmx12g dropseq.jar TagReadWithGeneExon I=0001.sub.bam O=0001.sub.GE.bam ANNOTATIONS_FILE=~/cellranger_references/refdata-cellranger-mm10-1.2.0/genes/genes.gtf TAG=GE ALLOW_MULTI_GENE_READS=true USE_STRAND_INFO=true VALIDATION_STRINGENCY=SILENT

samtools index 0001.sub.GE.bam

 

```


*** parameters ***

**INPUT=,I=** (required)  

The input SAM or BAM file to analyze 

**OUTPUT=,O=** (required)  

The output BAM, written with new Gene/Exon tag 

**TAG=** (required)  

The tag name to use.  When there are multiple genes, they will be comma seperated (Default value: GE)

**USE_STRAND_INFO=Boolean**

Use strand info to determine what gene to assign the read to. If this is on, reads can be assigned to a maximum one gene (Default value: true)

**ALLOW_MULTI_GENE_READS=Boolean**

Allow a read to span multiple genes.  If set to true, the gene name will be set to all of the gene/exons the read spans. In that case, the gene names will be comma separated (Default value: false)



### add read sequence and QV values to Nanopore SAM records  

SAMrecords are tagged with read sequence (US) and read quality (UQ). Is used for BC/UMI assignment and molecule consensus calculation.

QVs are currently just used to generate some QV stats during BC and UMI assignment -> adding QVs is optional.     

Fast, all jobs can be run on one compute node.  

command shown for BAM batch "0001.sub.bam"  

 
uses **AddBamReadSequenceTag (Scicelor.jar)**  


```bash

java -jar -Xmx12g sicelor.jar AddBamReadSequenceTag I=0001.sub.GE.bam O=0001.sub.GEUS.bam FASTQ=nanopore.fastq

samtools index 0001.sub.GEUS.bam

```

*** parameters ***

**INPUT=,I=** (required)  

The input SAM or BAM file to analyze 

**OUTPUT=,O=** (required)  

The output BAM, written with new Gene/Exon tag 

**FASTQ=** (required)  

The Nanopore .fastq file


<a id="bc-assignment"></a>

## 5) Nanopore Barcode/UMI assignment 

 

- Searches for cell barcodes and UMIs in genome matched Nanopore BAM files.

- Uses the barcode/UMI info of the Illumina short read data ( [IlluminaParser.jar](#parsing-illumina-data)) to guide barcode assignment.

- Will scan for poly(A) and adapter since positions are needed to extract potential cell barcode and UMI sequences.

 

Uses a "brute force approach".   

For barcode assignment the putative barcode sequence flanking the 10xGenomics adapter is extracted and matching barcodes for the same gene are searched in the Illumina data. If no match is found, all possible mutants (including indels) of the sequence are generated up to a defined maximal edit distance and tested for matching barcodes for the same gene in the Illumina data. If no match was found for the same gene or if the Nanopore SAM record is not annotated with a gene, the search is carried out for matching barcodes for the same genomic region (default window size 500 nt).

The maximal edit distance can be fixed or dynamic. Dynamic edit distance adapts the maximal edit distance to the complexity of the search set. 

**bcMaxEditDistances.xml** contains maximal edit distances allowed for different false assignment percentages.  

For UMI assignment, matching UMIs are only searched when a cell barcode was identified. Matching UMIs are searched for the same cell and the same gene or genomic region in the Illumina data. Just as the barcode edit distance, the UMI edit distance can be fixed or dynamically adjusted to the complexity of the search set (umiMaxEditDistances.xml). 

 

Found cell barcodes are added to the BC tag and UMI sequences to the U8 tag by default.  

Info on barcode and UMI position in read, edit distances and other info is written to various tags. See "config.xml" for more details and 

customization.

 

Paralellize, process batches generated in [(4)](#nanopore-tagging) on compute nodes if > 5 million reads. 

 

 

### Required files

 

* Nanopore_BC_UMI_finder.jar

* Libraries in the ./lib folder

* Config.xml: Config file to set various parameters.

* bcMaxEditDistances.xml and umiMaxEditDistances.xml: contain max edit distances for various barcode and UMI false assignment percentages respectively.

 

If no config files are found in the current path (working directory) , the software takes the config files from the directory where the application (jar) is.

 

### Usage

 

```bash  

java -jar -Xmx30g NanoporeBC_UMI_finder.jar  -i <Nanopore Bam, e.g. 0001.sub.GEUS.bam > 

-o <cell bc and UMI assigned output bam file>

-k <Illumina data previously parsed with IlluminaParser.jar>

--maxUMIfalseMatchPercent 2

--maxBCfalseMatchPercent 5

--logFile <path to log output file>

```

 

**Note:** The parameters above result in a effective barcode false assignment rate of 0.1% since subsequent UMI scan [eliminates most mis-assigned barcodes](#bc-misassignment).  

 

Highly multithreaded, will efficiently use all CPU cores at close to 100%.  

 

**RAM requirement:**  

Keeps the Illumina data and some info on parsed reads in RAM. 30 G is largely sufficient to scan a entire Promethion run with 50 - 100 million reads. A chunk with 20 million reads can typically be scanned with about 10 Gb of RAM.

We typically give most of or all the RAM available on the compute node since CPUs are used at 100% anyway.   

**Speed:**  

With 30 Gb attributed RAM, a 20 core Xeon node processes about 300,000 - 500,000 SAM records/hour when default settings are used.  

 

**Command line arguments override defaults in config xml.**

 

#### Required  

 
* **-i,--inFileNanopore**  

Nanopore input BAM file. The genome aligned Nanopore BAM input file with the read sequence in “US” tag and the gene name (if any) in the “GE” tag. All tags can be changed in “config.xml”.

* **-k,--inFile10x**  

10x Illumina File parsed by [IlluminaParser.jar](#parsing-illumina-data)

* **-y,--maxBCfalseMatchPercent**   

either this or –bcedit should  be provided  

Maximal percentage of false BC association. Will dynamically adjust BC edit distance depending on the numbers of cell barcodes to compare with. 

Uses data from simulation in file bcMaxEditDistances.xml (can be changes in config.xml) in current working directory and if not found there in application root overrides fixed BC edit distance parameter.  

<a id="bc-misassignment"></a>**Note:** The actual barcode false discovery rate is the product of barcode and UMI false discovery rate since a falsely assigned cell barcode will lead during subsequent UMI assignment to a search for matching UMIs among the UMIs associated with this false barcode --> UMI assignment will fail for the 

vast majority of wrong assigned barcodes.

e.g. --maxBCfalseMatchPercent 5  --maxUMIfalseMatchPercent 2 will lead to 0.05 * 0.02 = 0.1% false assigned barcodes amongst the barcode and UMI assigned 

reads and an effective barcode assignment accuracy of 99.9%. 

* **-b,--bcedit** either this or –maxBCfalseMatchPercent should be provided  

Fixed edit distance; max mismatches for cell barcode (fixed edit distance)

* **-j,--maxUMIfalseMatchPercent** either this or --umiedit should be provided  

Maximal percentage of false UMI association. Will dynamically adjustUMI edit distance depending on the numbers of umis to compare with. 

Uses data from similation in file umiMaxEditDistances.xml in current working directory and if not found there in application root. Overides fixed umi edit distance parameters.

* **-u,--umiedit**  either this or –--maxUMIfalseMatchPercent should be provided.   

Fixed edit distance; max errors (fixed edit distance) for umi

 

#### Optional (defaults can be modified in config xml)  

 

* **-o,--outfile**  

output bam file, if ommited will write file (inputfilename +"10xAttributes.bam" and file that contains only SAM entries with identified cellBC/UMI into <userhome>/<outdir> directory. Outdir defaults to “Nanopore_BC_UMIfinder”(set in config.xml).  

* **-l,--logFile**  

If not provided will create a default log with same name as outfile with “.log” appended in folder where output bam goes.  

* **-a,--polyawin**  

Window to search for poly(A/T) at both extremities, defaults to :100 (default in config.xml)

 

* **-n,--incrBCedit** (related to --bcedit)  

increase BC edit distance by one for genes expressed in less than 300 cells (defined in config.xml). Will only affect testing barcodes associated with gene/genomic region. Won't affect testing for all selected 10x barcodes or empty drop barcodes. Ignored unless fixed edit distance is used

* **-m,--incrumiedit**  

increase umi edit distance by one for genes with less than 50 umis (defined in config.xml) for gene in given cell. 

Ignored when dynamic edit 

distance is used (--maxUMIfalseMatchPercent).

* **-e,--randomBarcode; -f,--randomUMI**  

performs random Barcode or UMI Simulation. If a previously scanned and BC and UMI annotated file (the file with all entries not just the 

entries with found BC/UMIs) is rescanned with this option the max edit distance is the ED of the previously found match if any. Provide the barcode and UMI assigned BAM file (the file with all sam entries not just the BC/UMI assigned) as input BAM for the 

random barcode or umi scan (retrieves edit sdistance info for found BCs and UMIs there). 

* **-g,--ONTgene**  

2 char SAM attribute for gene name in Nanopore SAM. Default: GE (from config.xml)

* **-h,--help**  

* **-p,--adapterSeq** (default in config.xml)  

adapter sequence preceeeding UMI or barcode. Has to be in config file or here.

* **-q,--adapterMaxErr** (default in config.xml)  

max mismatches in adapter (Needleman alignment), default 3

* **-r,--tsMaxErr** (default in config.xml)  

max number of TSO mismatches, default: 3

* **-s,--tsoseq** (default in config.xml)  

TSO sequence, optional will be searched for and flagged if provided here or in “config.xml”, not used for read selection

**-t,--ncpu**  

n threads generated - if not provided will use all available CPU cores

**-w,--polyAlength** (default in config.xml)    

min polyA/T length, default 15.

* **-x,--polyATfraction**  (default in config.xml)   

fraction  A/T  in poly A/T tail, default: 0.75

* **-z,--edBCbailout** (default 2, in config.xml)    

mut cycles (ED) after which testing is aborted for Barcode when match was found. 

Avoids unnecessary testing and speeds up assignment but results in less info on secondary matches. E.g. if bail out at ED=2 and a match was found with ED=2 it won’t flag secondary matches at ED=4 in sam entry.

Defaults to always bail out at ED=2 if match was found. 

 

#### SAM tags in output file

 

SAM tags in output bam can be modified in the config xml. If entry for tag is commented out in XML it won’t be printed.  By default cell barcode is in BC tag and UMI in U8 tag.

Other info is added to the following tags by default:

 

- **Illumina data tags**    

 

**IG** If barcode was found for matching Gene in Illumina data, the gene name from the Illumina data is here  

 

- **polyA/T finding tags**  

 

**AX** Read too short, not analyzed  

**AZ** Pre-polyA/T sequence too short  

**A0** Neither polyA nor polyT  

**PE** End of polyA or polyT (1-based)  

**PI** Internal polyAT (evtl. adapter). Tag is set if internal poly(A/T)  found (potential chimeric read)  Format: PA(for polyA)PT(for polyT)<begin-end>AD(if adapter found,)<adapterbegin-adapterend> ------ entire thing is concatenated when multiple matches were found.    

                If "AD" is in string it is very likely that read is from a concatameric cDNA.  

 

- **Pre adapter finding checks and adapter Finding**  

 

**A2** Both polyA and polyT    

**AR** Read was reversed for scanning (poly A at 5')  

**AM**              Adapter not found  

**AT** Too many adapter mismatches  

**AB** Adapter start (1-based)  

**AE** Adapter end (1-based)  

 

- **TSO Finding**

 

**TN** TSO not found  

**TM**               TSO too many mismatches  

**TB** TSO begin (1-based)  

**TE** TSO end (1-based)   

 

- **Barcode Finding**  

 

**BN** Barcode not found  

**B0** Gene not found in Illumina data  

**B7** Genomic region not found in Illumina data  

**BF** Cell BC found for matching Illumina Cell BC for data for same gene in same cell  

**BR** Cell BC found for matching Illumina Cell BC for data for same genomic region in same cell  

**BT** Cell BC only found for a cell that does not have Illumina data for this gene or genomic region  

**B9** Cell BC only found in Illumina empty drops data  

**BG** No gene in Nanopore SAM  

**BS** Seq between adapter and polyAT too short to identify BC/UMI  

**BM**              More than one BC match with equal quality  

**B1** BC edit distance (mutation cycles used)  

**B2** BC edit distance second best match if any  (mutation cycles used)   

**BB** Barcode start (1-based)  

**BE** Barcode end (1-based)  

**BC** Cell BC sequence  

 

- **Umi Finding**  

 

**U9**               More than one match with equal quality  

**UT** Umi scan aborted: Post BC sequence too short  

**U1** UMI edit distance (mutation cycles)  

**U2** UMI edit distance (mutation cycles) of second best match (if any)  

**UB** UMI start (1 based)  

**UE** UMI end (one based)  

**U8** UMI sequence 

**UR** UMI was found after scan with reduced UMI length   

 

<a id="isoformmatrix"></a>

## 6) Transcript isoform expression quantification

 

**merge bam files**

Cell barcode and UMI assigned Bam files need to be merged.  

Requires: Picard-tools or Samtools


```bash  

java -jar -Xmx44g ~/picard-tools-1.119/MergeSamFiles.jar INPUT=0001.sub.GEUS10xAttributes.bam INPUT=0002.sub.GEUS10xAttributes.bam INPUT=... ASSUME_SORTED=true USE_THREADING=true TMP_DIR=/scratch/tmp/ MAX_RECORDS_IN_RAM=100000000 OUTPUT=GEUS10xAttributes.umifound.bam VALIDATION_STRINGENCY=SILENT

samtools index GEUS10xAttributes.umifound.bam

```


**isoform expression quantification**

**use IsoformMatrix (sicelore.jar)**

SAM records matching known genes were grouped by UMI and analyzed for matching Gencode v18 transcripts. SAM records with extensive non-matching sequences at either end, hard- or soft clipping > 150 nt (MAXCLIP parameter) are discarded. To assign a UMI to a Gencode transcript two strategies are available names STRICT and SOFT (METHOD paramater). Using STRICT strategy, a read is assigned to a given transcript when it recapitulates the full exon-exon junction layout authorizing a 2 bases margin (DELTA parameter) of added or lacking sequences at exon boundaries, to allow for indels at exon junctions and imprecise mapping by Minimap2. For each UMI, we analyze all its reads and assign the UMI the Gencode transcript supported by the majority of the reads. 

In case of an equal number of reads support two different Gencode transcript, the UMI is considered as ambiguous and assigned/not assigned to an isoform (AMBIGUOUS_ASSIGN parameter). SOFT strategy allows to assign isoform to an UMI that do not recapitulate the full exon-exon structure of the isoform. Some reads corresponding to long RNAs are incomplete (lacking 5') due to limitations of the 10xGenomics library preparation. Some reads lack the 3' end since the oligod(T) priming and reverse transcription starts at room temperature (low stringency) resulting in priming at internal A-rich sequences of the RNA. We require in this mode the visualisation of an exon-exon junction specific to an isoform of the gene model processed. Briefly for a UMI matching Gene X, all gencode transcript of a gene model for the gene X are parsed, each unique exon-exon junction is associated to the list of isoforms carrying the junction. All the reads for the UMI are processed, for each exon-exon junction, a score of 1 is added to the list of isoforms carrying the junction. At the end of the process, the isoform with the highest score is selected as the isoform for the UMI. If more than one Gencode transcript obtain the highest score, the UMI is considered as ambiguous and not assigned or asssigned to one of the ùatching isoform (AMBIGUOUS_ASSIGN=true).


**parameters**

**INPUT=**  (required)  

GEUS10xAttributes.umifound.bam file containing cellBC (BC tag) and UMI (U8 tag) assigned Nanopore SAM entries with the gene mame in the IG tag. SAMrecords lacking any of those 3 required fields are discarded from further analysis.   

**CSV=** (required)  

.csv file listing, one per line, the cell barcodes that need to be quantified (e.g. brain951.barcodes.tsv in github /barcodes/ directory)  

**REFFLAT=** (required)  

.txt file describing the gene model for the organism concerned (e.g. refFlat_gencode.vM18.txt in github /refFlat/ directory) 

**PREFIX=** (required)  

prefix used for output file names

**DELTA=** (required)  

the number of extra or lacking bases allowed at exon-exon junctions (default = 2)  

**MAXCLIP=** (required)  

the maximal number of extensive non-matching sequences at either end, hard- or soft clipping to call the read as chimerix and discards it (default = 150)  

**METHOD=** (required)  

STRICT for full exon-exon structure required for assignation

SOFT for a more lenient way of assignation requiring observation of an isoform specific exon-exon junction 

**AMBIGUOUS_ASSIGN=** (required)  

Wether or not to assign a UMI that is ambiguous (>1 isoforms of the gene model fit its exon-exon structure) (default=false)

**ISOBAM=** (required)  

Wether or not to output a BAM file containg a flag (IT) with the transcriptID called during this process (default=false)


**output files**

**PREFIX**_cells_metrics.txt

cell by cell metrics (number of reads, number of UMIs, %age of molecules isoform assigned)

**PREFIX**_genes_metrics.txt

gene by gene metrics (total molecules, %age of molecules isoforms assigned)

**PREFIX**_isoforms_metrics.txt

isoform by isoform metrics (total molecules)

**PREFIX**_genes_matrix.txt

gene level number of molecules per cell barcode count matrix for subsequent statistical analysis

**PREFIX**_isoforms_matrix.txt

isoform level number of molecules per cell barcode count matrix for subsequent statistical analysis

 
#### Example


```bash

java -jar -Xmx44g sicelor.jar IsoformMatrix I=GEUS10xAttributes.umifound.bam REFFLAT=refFlat_gencode.vM18.txt CSV=10xgenomics.barcodes.csv DELTA=2 MAXCLIP=150 METHOD=SOFT AMBIGUOUS_ASSIGN=false PREFIX=sicelore

```

<a id="computeConsensus"></a>

## 7) Compute consensus sequences

 
**use ComputeConsensus (sicelore.jar)**

The pipeline allows to compute the consensus sequence for each molecule identified in the INPUT .bam file.

First steps is similar to the IsoformMatrix pipeline except that the cDNA sequence defined as (tsoEnd(TE tag) ... umiEnd(UE tag)) is loaded so that it can be used for consensus sequence computation. 

Briefly, each molecule are processed as follow depending the number of reads the molecule behave: (i)case of a 1-read molecule, the consensus sequence is set to the 

read cDNA sequence; (ii) case of a 2-reads molecule, the consensus sequence is set as the cDNA sequence of the best quality read according

to the minimal "de" minimap2 SAMrecord tag value; (iii) Case of a multi-reads molecule (i.e. > 2), a consensus sequence is set

using poaV2 multiple alignment using all reads for the molecule. The consensus sequence is then "racon" polished. 

Using a 20 cores compute nodes for muti-threading, and depending on the sequencing depth inducing a low/high number of multi-reads molecules, the speed of consensus sequence computation is dependent of the sequencing depth but can be estimated at 200k UMIs/ hour.

For time calculation optimization, this step should be parrallelized, for instance on a per chromosome basis, and dispense on a calcul cluster.


### splitting bam by chromosomes  


```bash  

@jobs = ('1','2','X','MT','3','4',..all chromosomes..);

for($i=0; $i<@jobs; $i++){

    samtools view -Sb GEUS10xAttributes.umifound.bam ".$jobs[$i]." -o GEUS10xAttributes.umifound.chr".$jobs[$i].".bam

    samtools index GEUS10xAttributes.umifound.chr".$jobs[$i].".bam

}

```

 

### ComputeConsensus pipeline

**parameters**

**INPUT=**  (required)  

GEUS10xAttributes.umifound.chr1.bam file containing cellBC (BC tag) and UMI (U8 tag) assigned Nanopore SAM entries with the gene mame in the IG tag. SAMrecords lacking any of those 3 required fields are discarded from further analysis.   

**OUTPUT=,O=**  (required)  

Consensus sequence in fasta format.

**THREADS=,T=** (required)  

Number of threads for multi-threading (typically number of cores of compute node)

**TMPDIR=** (required)  

Temporary directory

**POAPATH=** (required)  

poaV2 path to directory where to find executable (/share/apps/bin/)

**RACONPATH=** (required)  

Racon path to directory where to find executable (/share/apps/bin/)

**MINIMAP2PATH=** (required)  

Minimap2 path to directory where to find executable (/share/apps/bin/)


code below is for chromosome 1, repeat for all chromosomes 

```bash

java -jar -Xmx22g sicelor.jar ComputeConsensus I=GEUS10xAttributes.umifound.chr1.bam O=molecules.chr1.fa TMPDIR=/tmp/ POAPATH=/share/apps/local/bio-pipeline/poaV2/ RACONPATH=/share/apps/local/racon/bin/ MINIMAP2PATH=/share/apps/local/minimap2/

```

<a id="molecule-consensus-mapping"></a>

## 8) Re-map consensus sequences to reference genome  

If compute consensus has been split per chromosome, you need to deduplicate molecules for reads mapping to several genomic position such as reads coming from genes having pseudogenes.

First you need to concatenate fasta file for all chromosomes then use ***DeduplicateMolecule*** pipeline (sicelore.jar)

```bash

cat */molecules.chr*.fa > molecules.fa

java -jar -Xmx22g sicelor.jar DeduplicateMolecule I=molecule.fa O=deduplicate.fa

```

Molecule consensus sequences can then be mapped back to the reference genome to generate a molecule .bam file for further analysis such as SNP calling or identification of RNA editing.  

```bash

minimap2 -ax splice -uf --MD -N 100 --sam-hit-only -t 20 --junc-bed gencode.vM18.primary_assembly.annotation.bed $BUILD.mmi molecules.fa > molecules.sam

samtools view -Sb molecules.sam -o unsorted.bam

samtools sort unsorted.bam -o molecules.bam

samtools index molecules.bam

```

CellBC / UMI tags can be added to the SAM records using the ***AddBamMoleculeTags*** pipeline (sicelore.jar).

``` bash  

java -jar -Xmx22g sicelor.jar AddBamMoleculeTags I=molecule.bam O=molecule.tags.bam

samtools index molecules.tags.bam

```
