---
title: "E18 mouse brain, isoforms differentially expressed"
author: "K.Lebrigand, R.Waldmann"
output:
  BiocStyle::html_document:
    code_folding: hide
    number_sections: yes
    toc: yes  
---

# Loading illu1141c.rds

```{r message=FALSE, warning=FALSE, echo=FALSE, eval=T, fig.height=6, fig.width=8}

source("00.import.R")

illu1141 <- readRDS("illu1141c.rds")
DimPlot(object = illu1141, reduction = "tsne", label = FALSE, pt.size = 2)

DefaultAssay(object = illu1141) <- "RNA"
g.markers <- FindAllMarkers(object = illu1141, only.pos = TRUE, min.pct = 0.1, logfc.threshold = 0.25)
write.table(g.markers, file="output/markers.genes.csv", sep=",")

DefaultAssay(object = illu1141) <- "ISO"
iso.markers <- FindAllMarkers(object = illu1141, only.pos = TRUE, min.pct = 0.1, logfc.threshold = 0.25)
#dim(iso.markers)

iso.markers$geneId <- sapply(strsplit(rownames(iso.markers), "-ENS"), `[`, 1)
iso.markers$transcriptId <- paste("ENS", sapply(strsplit(rownames(iso.markers), "-ENS"), `[`, 2), sep="")

```

# Extracting tables 

```{r message=FALSE, warning=FALSE, echo=FALSE, eval=T, fig.height=6, fig.width=26}

iso.markers$ISO.only <- "no"
iso.markers[which(!(iso.markers$geneId %in% g.markers$gene)),]$ISO.only <- "yes"
table(iso.markers$ISO.only)

g.genes <- g.markers$gene
g.markers$min.pval.adj <- 1
for(i in 1:length(g.genes)){
  g.markers$min.pval.adj[i] <- min(g.markers$p_val_adj[g.markers$gene == g.genes[i]])
}

iso.genes <- iso.markers$geneId
iso.markers$min.pval.adj <- 1
for(i in 1:length(iso.genes)){
  iso.markers$min.pval.adj[i] <- min(iso.markers$p_val_adj[iso.markers$geneId == iso.genes[i]])
}

iso.markers$min.pval.adj.genes <- "#N/A"
for(i in 1:length(iso.genes)){
  #print (iso.genes[i])
  if(iso.genes[i] %in% g.markers$gene){
    iso.markers$min.pval.adj.genes[i] <- g.markers$min.pval.adj[g.markers$gene == iso.genes[i]]
  }
}

#head(iso.markers)
write.table(iso.markers, file="output/markers.isoforms.csv", sep=",")

```

# Differentially expressed isoforms

```{r message=FALSE, warning=FALSE, echo=FALSE, eval=T, fig.height=6, fig.width=26}

DefaultAssay(object = illu1141) <- "ISO"
clusters <- c("cycling radial glia","radial glia","intermediate progenitor", "imature Glutamatergic", "mature Glutamatergic","imature GABAergic","mature GABAergic", "Cajal-Retzius")

total <- data.frame()
for (i in (1:7)){
  k <- i+1
  for (j in (k:8)){
    if(i != j){
      print(paste(i, " ", j, " ",clusters[i], " vs ", clusters[j], sep=""))
      
      markers <- FindMarkers(object = illu1141, ident.1=clusters[i], ident.2=clusters[j], min.pct = 0.1, thresh.use = 0.25)
      markers$cluster <- clusters[j]
      markers$contrast <- paste(clusters[i], "vs", clusters[j], sep=" ")
      markers[which(markers$avg_logFC>0),]$cluster <- clusters[i]
      markers$geneId <- sapply(strsplit(rownames(markers), "-ENS"), `[`, 1)
      markers$transcriptId <- paste("ENS", sapply(strsplit(rownames(markers), "-ENS"), `[`, 2), sep="")
      #total <- markers[which(markers$geneId == "------"),]
      all.genes <- unique(markers$geneId)
      for (k in (1:length(all.genes))){
         sub <- markers[which(markers$geneId == all.genes[k]),]
         nb.clusters <- unique(sub$cluster)
         nb.transcripts <- unique(sub$transcriptId)
        
         if(length(nb.clusters) > 1 & length(nb.transcripts) > 1){
           total <- rbind(total, sub)
         }
      }
      print (dim(total))
      print (length(unique(total$geneId)))
    }
  }
}

length(unique(total$geneId))
head(total)

write.table(total, file="output/markers.isoforms.only.csv", sep=",")

```

# Session Info

```{r sessinf}
sessionInfo()
```




